File: src/servicebroker-npm/src/IpcRelayServiceBroker.ts
import CancellationToken from 'cancellationtoken'
import { PIPE_NAME_PREFIX, RemoteServiceConnections } from './constants'
import EventEmitter = require('events')
import { RemoteServiceConnectionInfo } from './RemoteServiceConnectionInfo'
import { ServiceActivationOptions } from './ServiceActivationOptions'
import { ServiceBrokerClientMetadata } from './ServiceBrokerClientMetadata'
import { ServiceMoniker } from './ServiceMoniker'
import { IDisposable } from './IDisposable'
import { IRemoteServiceBroker } from './IRemoteServiceBroker'
import { ServiceBrokerEmitter, IServiceBroker } from './IServiceBroker'
import { randomUUID } from 'crypto'
import { createServer, Server } from 'net'
import { BrokeredServicesChangedArgs } from './BrokeredServicesChangedArgs'
import { RpcEventServer } from './ServiceRpcDescriptor'
import path = require('path')
import { chmodSync, mkdtempSync, rmSync } from 'fs'
import { tmpdir } from 'os'

/**
 * An IRemoteServiceBroker which proffers all services from another IServiceBroker
 * over named pipes on Windows or Unix domain sockets on other operating systems.
 */
export class IpcRelayServiceBroker extends (EventEmitter as new () => ServiceBrokerEmitter) implements IRemoteServiceBroker, RpcEventServer, IDisposable {
	private static readonly _rpcEventNames = Object.freeze(['availabilityChanged'])
	readonly rpcEventNames = IpcRelayServiceBroker._rpcEventNames
	public readonly completion: Promise<void>
	private readonly channelsOfferedToClient: { [Key: string]: Server } = {}
	private disposed: (() => void) | undefined

	constructor(private readonly serviceBroker: IServiceBroker) {
		super()
		serviceBroker.on('availabilityChanged', this.onAvailabilityChanged.bind(this))
		this.completion = new Promise<void>(resolve => (this.disposed = resolve))
	}

	handshake(clientMetadata: ServiceBrokerClientMetadata, cancellationToken?: CancellationToken): Promise<void> {
		if (!RemoteServiceConnections.contains(clientMetadata.supportedConnections, RemoteServiceConnections.IpcPipe)) {
			throw new Error('The client must support IpcPipe to use this service broker.')
		}

		return Promise.resolve()
	}
	async requestServiceChannel(
		serviceMoniker: ServiceMoniker,
		options?: ServiceActivationOptions,
		cancellationToken?: CancellationToken
	): Promise<RemoteServiceConnectionInfo> {
		const pipe = await this.serviceBroker.getPipe(serviceMoniker, options, cancellationToken)
		if (!pipe) {
			return {}
		}

		const requestId = randomUUID()
		const pipeDirectory = process.platform === 'win32' ? undefined : IpcRelayServiceBroker.createPipeDirectory()
		const pipeName = pipeDirectory ? path.join(pipeDirectory, randomUUID()) : path.join(PIPE_NAME_PREFIX, randomUUID())

		const server = createServer()
		server.once('connection', serverPipe => {
			// Drop the entry from our map once the connection has been made.
			delete this.channelsOfferedToClient[requestId]
			server.close()

			serverPipe.pipe(pipe)
			pipe.pipe(serverPipe)
		})
		if (pipeDirectory) {
			server.once('close', () => rmSync(pipeDirectory, { recursive: true, force: true }))
		}
		this.channelsOfferedToClient[requestId] = server

		try {
			await IpcRelayServiceBroker.listen(server, pipeName)
		} catch (error) {
			delete this.channelsOfferedToClient[requestId]
			if (server.listening) {
				server.close()
			}
			if (pipeDirectory) {
				rmSync(pipeDirectory, { recursive: true, force: true })
			}

			throw error
		}

		return {
			requestId,
			pipeName,
		}
	}
	cancelServiceRequest(serviceRequestId: string, cancellationToken?: CancellationToken): Promise<void> {
		const server = this.channelsOfferedToClient[serviceRequestId]
		if (server) {
			delete this.channelsOfferedToClient[serviceRequestId]
			server.close()
			return Promise.resolve()
		} else {
			return Promise.reject('Request to cancel a channel that is not awaiting acceptance.')
		}
	}
	dispose() {
		this.serviceBroker.off('availabilityChanged', this.onAvailabilityChanged.bind(this))
		for (const requestId in this.channelsOfferedToClient) {
			const server = this.channelsOfferedToClient[requestId]
			server.close()
			delete this.channelsOfferedToClient[requestId]
		}

		if (this.disposed) {
			this.disposed()
		}
	}

	private onAvailabilityChanged(args: BrokeredServicesChangedArgs) {
		this.emit('availabilityChanged', args)
	}

	private static createPipeDirectory(): string {
		const pipeDirectory = mkdtempSync(path.join(tmpdir(), 'servicehub-ipc-'))
		chmodSync(pipeDirectory, 0o700)
		return pipeDirectory
	}

	private static listen(server: Server, pipeName: string): Promise<void> {
		return new Promise<void>((resolve, reject) => {
			const onError = (error: Error) => {
				server.off('listening', onListening)
				reject(error)
			}
			const onListening = () => {
				server.off('error', onError)
				resolve()
			}

			server.once('error', onError)
			server.once('listening', onListening)
			server.listen(pipeName)
		})
	}
}

File: src/Microsoft.ServiceHub.Framework/IpcRelayServiceBroker.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System.Collections.Immutable;
using System.Diagnostics;
using System.IO.Pipelines;
using Microsoft.VisualStudio.Threading;
using IAsyncDisposable = System.IAsyncDisposable;

namespace Microsoft.ServiceHub.Framework;

/// <summary>
/// An <see cref="IRemoteServiceBroker"/> which proffers all services from another <see cref="IServiceBroker"/>
/// over named pipes on Windows or Unix domain sockets on other operating systems.
/// </summary>
public class IpcRelayServiceBroker : IRemoteServiceBroker, IDisposable
{
	private readonly IServiceBroker serviceBroker;

	/// <summary>
	/// An event to set upon disposal.
	/// </summary>
	private readonly AsyncManualResetEvent disposedEvent = new AsyncManualResetEvent();

	private ImmutableDictionary<Guid, IAsyncDisposable> remoteServiceRequests = ImmutableDictionary<Guid, IAsyncDisposable>.Empty;

	/// <summary>
	/// Initializes a new instance of the <see cref="IpcRelayServiceBroker"/> class.
	/// </summary>
	/// <param name="serviceBroker">The service broker whose services are to be exposed.</param>
	public IpcRelayServiceBroker(IServiceBroker serviceBroker)
	{
		Requires.NotNull(serviceBroker, nameof(serviceBroker));
		this.serviceBroker = serviceBroker;
	}

	/// <inheritdoc/>
	public event EventHandler<BrokeredServicesChangedEventArgs> AvailabilityChanged
	{
		add => this.serviceBroker.AvailabilityChanged += value;
		remove => this.serviceBroker.AvailabilityChanged -= value;
	}

	/// <summary>
	/// Gets the logging mechanism.
	/// </summary>
	public TraceSource? TraceSource { get; init; }

	/// <summary>
	/// Gets a <see cref="Task"/> that completes when this instance is disposed of.
	/// </summary>
	/// <remarks>
	/// This event will occur when the client disconnects from the relay,
	/// if the RPC library is configured to dispose target objects at that time.
	/// </remarks>
	public Task Completion => this.disposedEvent.WaitAsync();

	/// <inheritdoc />
	public Task HandshakeAsync(ServiceBrokerClientMetadata clientMetadata, CancellationToken cancellationToken = default)
	{
		if (!clientMetadata.SupportedConnections.HasFlag(RemoteServiceConnections.IpcPipe))
		{
			// Only support pipes.
			throw new NotSupportedException();
		}

		return Task.CompletedTask;
	}

	/// <inheritdoc />
	public async Task<RemoteServiceConnectionInfo> RequestServiceChannelAsync(ServiceMoniker serviceMoniker, ServiceActivationOptions serviceActivationOptions, CancellationToken cancellationToken = default)
	{
		DisposableBag faultOrCancelBag = new();
		try
		{
			IDuplexPipe? servicePipe = await this.serviceBroker.GetPipeAsync(serviceMoniker, serviceActivationOptions, cancellationToken).ConfigureAwait(false);
			if (servicePipe == null)
			{
				return default;
			}

			faultOrCancelBag.AddDisposable(new DisposeAction(() =>
			{
				servicePipe.Input?.Complete();
				servicePipe.Output?.Complete();
			}));

			var requestId = Guid.NewGuid();

			IIpcServer server = ServerFactory.Create(
				stream => this.HandleIncomingConnectionAsync(stream, requestId, servicePipe),
				new ServerFactory.ServerOptions { TraceSource = this.TraceSource });
			Assumes.True(faultOrCancelBag.TryAddDisposable(server));

			ImmutableInterlocked.TryAdd(ref this.remoteServiceRequests, requestId, faultOrCancelBag);

			return new RemoteServiceConnectionInfo
			{
				RequestId = requestId,
				PipeName = server.Name,
			};
		}
		catch
		{
			await faultOrCancelBag.DisposeAsync().ConfigureAwait(false);
			throw;
		}
	}

	/// <inheritdoc />
	public async Task CancelServiceRequestAsync(Guid serviceRequestId)
	{
		if (ImmutableInterlocked.TryRemove(ref this.remoteServiceRequests, serviceRequestId, out IAsyncDisposable? disposable))
		{
			await disposable.DisposeAsync().ConfigureAwait(false);
		}
	}

	/// <inheritdoc/>
	public void Dispose()
	{
		this.Dispose(true);
		GC.SuppressFinalize(this);
	}

	/// <summary>
	/// Disposes managed and unmanaged resources owned by this instance.
	/// </summary>
	/// <param name="disposing"><see langword="true" /> if this object is being disposed; <see langword="false" /> if it is being finalized.</param>
	protected virtual void Dispose(bool disposing)
	{
		if (disposing)
		{
			this.disposedEvent.Set();
		}
	}

	private async Task HandleIncomingConnectionAsync(Stream stream, Guid requestId, IDuplexPipe servicePipe)
	{
		// Once a connection is made (or fails), it is no longer cancelable.
		ImmutableInterlocked.TryRemove(ref this.remoteServiceRequests, requestId, out IAsyncDisposable _);

		// Link the two pipes so that all incoming/outgoing calls get forwarded
		await Task.WhenAll(
			stream.CopyToAsync(servicePipe.Output),
			servicePipe.Input.CopyToAsync(stream)).ConfigureAwait(false);
	}
}
